File: client/test/GhostBeInterceptorTest.kt
@file:OptIn(okhttp3.ExperimentalOkHttpApi::class)

package dev.yellobytes.ghostbe.client

import kotlin.test.Test
import kotlin.test.AfterTest
import kotlin.test.assertEquals
import mockwebserver3.MockResponse
import mockwebserver3.MockWebServer
import okhttp3.OkHttpClient
import okhttp3.Request

class GhostBeInterceptorTest {
    private val ghostBe = MockWebServer()
    private val realBackend = MockWebServer()

    @AfterTest
    fun tearDown() {
        ghostBe.shutdown()
        realBackend.shutdown()
    }

    private fun clientPointedAt(baseUrl: String) = OkHttpClient.Builder()
        .addInterceptor(GhostBeInterceptor(baseUrl = baseUrl))
        .build()

    @Test
    fun `returns the mocked response when ghost-be intercepts`() {
        ghostBe.start()
        ghostBe.enqueue(
            MockResponse.Builder()
                .code(200)
                .body("""{"intercept":true,"status":201,"headers":{"Content-Type":"application/json"},"body":"eyJvayI6dHJ1ZX0="}""")
                .build()
        )

        val client = clientPointedAt(ghostBe.url("/").toString())
        val request = Request.Builder().url(realBackend.url("/v1/users/42")).build()
        val response = client.newCall(request).execute()

        assertEquals(201, response.code)
        assertEquals("""{"ok":true}""", response.body!!.string())
        assertEquals(0, realBackend.requestCount) // never called the real backend
    }

    @Test
    fun `calls the real backend when ghost-be passes through`() {
        ghostBe.start()
        ghostBe.enqueue(MockResponse.Builder().code(200).body("""{"intercept":false}""").build())
        realBackend.start()
        realBackend.enqueue(MockResponse.Builder().code(200).body("real response").build())

        val client = clientPointedAt(ghostBe.url("/").toString())
        val request = Request.Builder().url(realBackend.url("/v1/users/42")).build()
        val response = client.newCall(request).execute()

        assertEquals("real response", response.body!!.string())
        assertEquals(1, realBackend.requestCount)
    }

    @Test
    fun `calls the real backend when ghost-be is unreachable`() {
        // ghost-be is never started, so its url points at a closed port.
        realBackend.start()
        realBackend.enqueue(MockResponse.Builder().code(200).body("real response").build())

        val client = clientPointedAt("http://127.0.0.1:1") // port 1 refuses connections
        val request = Request.Builder().url(realBackend.url("/v1/users/42")).build()
        val response = client.newCall(request).execute()

        assertEquals("real response", response.body!!.string())
        assertEquals(1, realBackend.requestCount)
    }
}

File: client-flutter/test/ghost_be_interceptor_test.dart
import 'dart:convert';
import 'dart:io';
import 'package:dio/dio.dart';
import 'package:test/test.dart';
import 'package:ghost_be/ghost_be.dart';

Future<HttpServer> _stub(String responseJson) async {
  final server = await HttpServer.bind('127.0.0.1', 0);
  server.listen((request) async {
    request.response.headers.contentType = ContentType.json;
    request.response.write(responseJson);
    await request.response.close();
  });
  return server;
}

void main() {
  tearDown(GhostBe.clearCapturedServer);

  test('resolves the mocked response without calling the real backend', () async {
    final ghostBe = await _stub(jsonEncode({
      'intercept': true,
      'status': 201,
      'headers': {'Content-Type': 'application/json'},
      'body': base64Encode(utf8.encode('{"ok":true}')),
    }));
    addTearDown(() => ghostBe.close(force: true));

    final dio = Dio()..interceptors.add(GhostBeInterceptor(baseUrl: 'http://127.0.0.1:${ghostBe.port}'));
    final response = await dio.get<List<int>>(
      'http://example.invalid/v1/users/42',
      options: Options(responseType: ResponseType.bytes),
    );

    expect(response.statusCode, 201);
    expect(utf8.decode(response.data!), '{"ok":true}');
  });

  test('falls through to the real request when ghost-be says passthrough', () async {
    final ghostBe = await _stub(jsonEncode({'intercept': false}));
    addTearDown(() => ghostBe.close(force: true));
    final realBackend = await _stub(jsonEncode({'real': 'backend'}));
    addTearDown(() => realBackend.close(force: true));

    final dio = Dio()..interceptors.add(GhostBeInterceptor(baseUrl: 'http://127.0.0.1:${ghostBe.port}'));
    final response = await dio.get<String>('http://127.0.0.1:${realBackend.port}/v1/users/42');

    expect(response.statusCode, 200);
    expect(response.data, jsonEncode({'real': 'backend'}));
  });

  test('falls through when ghost-be is unreachable', () async {
    final realBackend = await _stub(jsonEncode({'real': 'backend'}));
    addTearDown(() => realBackend.close(force: true));

    // Port 1 refuses connections -- nothing is listening there.
    final dio = Dio()..interceptors.add(GhostBeInterceptor(baseUrl: 'http://127.0.0.1:1'));
    final response = await dio.get<String>('http://127.0.0.1:${realBackend.port}/v1/users/42');

    expect(response.statusCode, 200);
    expect(response.data, jsonEncode({'real': 'backend'}));
  });

  test('deep-link captured server overrides the interceptor\'s configured baseUrl', () async {
    final ghostBe = await _stub(jsonEncode({
      'intercept': true,
      'status': 201,
      'headers': {'Content-Type': 'application/json'},
      'body': base64Encode(utf8.encode('{"ok":true}')),
    }));
    addTearDown(() => ghostBe.close(force: true));

    // Constructed pointing at a dead port -- the deep link is what actually gets used.
    final dio = Dio()..interceptors.add(GhostBeInterceptor(baseUrl: 'http://127.0.0.1:1'));
    GhostBe.captureFromUri(Uri.parse('myapp://open?ghostBe=127.0.0.1:${ghostBe.port}'));

    final response = await dio.get<List<int>>(
      'http://example.invalid/v1/users/42',
      options: Options(responseType: ResponseType.bytes),
    );

    expect(response.statusCode, 201);
    expect(utf8.decode(response.data!), '{"ok":true}');
  });

  test('clearCapturedServer falls back to the interceptor\'s configured baseUrl', () async {
    final ghostBe = await _stub(jsonEncode({
      'intercept': true,
      'status': 201,
      'headers': {'Content-Type': 'application/json'},
      'body': base64Encode(utf8.encode('{"ok":true}')),
    }));
    addTearDown(() => ghostBe.close(force: true));

    GhostBe.captureFromUri(Uri.parse('myapp://open?ghostBe=127.0.0.1:1')); // dead port
    GhostBe.clearCapturedServer();

    final dio = Dio()..interceptors.add(GhostBeInterceptor(baseUrl: 'http://127.0.0.1:${ghostBe.port}'));
    final response = await dio.get<List<int>>(
      'http://example.invalid/v1/users/42',
      options: Options(responseType: ResponseType.bytes),
    );

    expect(response.statusCode, 201);
    expect(utf8.decode(response.data!), '{"ok":true}');
  });
}
